solutions/USACO jan17-cownav.mdx

```diff
@@ -38,6 +38,157 @@ Don't forget that once Bessie reaches the goal, she will ignore further commands
 
 ### Implementation
 
-<!-- I really don't want to write Cow Navigation. Anyone have code? -->
 
-<IncompleteSection />
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+//created by Oleksandr Gorpynich
+
+//This implementation also keeps track of whether cow 1 (facing left) and cow 2 (facing up) 
+//have reached the end at some point. This adds a factor of 4 to the solution however it still
+//runs fast enough
+#pragma GCC optimize("O2")
+#include <fstream>
+#include <string>
+#include <iostream>
+#include <bitset>
+#include <math.h>
+#include <string>
+#include <algorithm>
+#include <assert.h>
+#include<bits/stdc++.h>
+#include <vector>
+#include <queue>
+#include<stdio.h>
+#include<ctype.h>
+#define ll long long
+using namespace std;
+
+int n;
+
+struct state{
+    int x1;
+    int y1;
+    int dir1;
+    int x2;
+    int y2;
+    int dir2;
+    int finished1;
+    int finished2;
+    int dist;
+};
+char arr[21][21];
+int visited[21][21][4][21][21][4][2][2];
+//x1, y1, dir1, x2, y2, dir2, finished1, finished2
+int dx[4] = {-1, 0, 1, 0};
+int dy[4] = {0, 1, 0, -1};
+//up, left, down, right
+//test if a cell is possible to visit
+bool inside(int x, int y){
+    if (x >= 0 && x < n && y >= 0 && y < n){
+        if (arr[x][y] != 'H'){
+            return true;
+        }
+    }
+    return false;
+}
+//set a state as visited
+bool setvisited(state cur){
+    visited[cur.x1][cur.y1][cur.dir1][cur.x2][cur.y2][cur.dir2][cur.finished1][cur.finished2] = 1;
+}
+//check if a state has been visited
+bool free(state cur){
+    if (visited[cur.x1][cur.y1][cur.dir1][cur.x2][cur.y2][cur.dir2][cur.finished1][cur.finished2] == -1){
+        return true;
+    }
+    return false;
+}
+int main(){
+    ifstream in("cownav.in");
+    ofstream out("cownav.out");
+    in >> n;
+    for (int x = 0; x < n; x++){
+        for (int y = 0; y < n; y++){
+            in >> arr[x][y];
+        }
+    }
+    for (int x = 0; x < n; x++){
+        for (int y = 0; y < n; y++){
+            for (int d1 = 0; d1 < 4; d1++){
+                for (int z = 0; z < n; z++){
+                    for (int b = 0; b < n; b++){
+                        for (int d2 = 0; d2 < 4; d2++){
+                            visited[x][y][d1][z][b][d2][0][0] = -1;
+                            visited[x][y][d1][z][b][d2][1][0] = -1;
+                            visited[x][y][d1][z][b][d2][0][1] = -1;
+                            visited[x][y][d1][z][b][d2][1][1] = -1;
+                        }
+                    }
+                }
+            }
+        }
+    }
```

Comment:
This big chunk can be simply
`memset(visited, -1, sizeof visited);`
or you can use [std::fill](http://www.cplusplus.com/reference/algorithm/fill/)